nixos/borgbackup: Borgbackup retry

toonn opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

I was lacking an option to specify borgbackup jobs should restart if
they failed. My remote borg repo isn't always available so setting up
weekly backups wasn't very useful. The backups failed every time so my
files were never backed up. Specifying `Restart=on-failure` and
`RestartSec=X min` should prevent spurious failures from ruining your
day when it comes time to restore your non-existent backups : )

###### Things done

I've run this patch (applied at nixos-19.09) for quite a while now and it's worked like a charm.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

toonn committed:
nixos/borgbackup: Add restartSec option

I was lacking an option to specify borgbackup jobs should restart if
they failed. My remote borg repo isn't always available so setting up
weekly backups wasn't very useful. The backups failed every time so my
files were never backed up. Specifying `Restart=on-failure` and
`RestartSec=X min` should prevent spurious failures from ruining your
day when it comes time to restore your non-existent backups : )

Might need a Persistent timer for more reliability, in case system
happens to be powered down every day at the time of backup.

toonn committed:
nixos/borgbackup: Add persistent and randomizedDelay

A persistent timer is useful if your machine is regularly turned off
around scheduled backup times.

It's useful to randomly delay the backup timers especially when they are
persistent because they tend to bunch together when powering up your
machine.

I've renamed the `persistentTimer` option to be in line with other NixOS
options like `nix.gc.persistent` and `systemd.timer` terminology.

I've renamed `mkBackupTimers` to the singular because the plural is
deceptive, the function only ever generates a config for a single timer.
I also alphabetized the attrsets here and for the options, excepting
`type` as seems to be the convention.

bjornfor commented:
Related: https://github.com/NixOS/nixpkgs/pull/85696

dotlambda commented:
I suggest you set a shorter interval instead and use `preHook` to check how old the last backup is. If it's less than a week old, just call `exit`.

toonn commented:
@dotlambda, why? That sounds like a worse version of using the appropriate systemd features? Systemd knows whether or not the unit ran successfully so it doesn't require any superfluous checking of the remote repository.

mweinelt commented:
> use `preHook` to check how old the last backup is

Is that even possible in an append-only scenario?

toonn commented:
Would appreciate a review on this.

Lassulus commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -277,6 +289,40 @@ in {
+            description = ''
+              Backups are delayed by a uniformly random amount of time up to
+              the value. Useful to stagger remote backups on a weak network
+              connection.
```
maybe a reference to the used time format would be nice

toonn replied:
I repeated the reference to the man page from the `startAt` option for the new options that expect time spans and rebased. I guess I figured the other reference was close enough before but from a distance it is nicer the reference is right there, yeah.

toonn committed:
nixos/borgbackup: Improve documentation for the new options

aanderse reviewed (approved):
From a high level this change LGTM... but I don't have `borgbackup` experience. What do we need to do to get this merged?

ping @Lassulus @mweinelt

:heart: @toonn

toonn commented:
@dotlambda, your scheduled end of the week (*very* end of the week) ping : )

dotlambda commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -279,6 +291,46 @@ in {
+              (in seconds) or a time span value, e.g.,
+              <literal>"5min 20s"</literal>.
+              If you do not want a failed backup to restart automatically, use
+              <literal>null</literal>.  '';
```
```suggestion
              <literal>null</literal>.
            '';
```

dotlambda commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -279,6 +291,46 @@ in {
+          };
+
+          restartSec = mkOption {
+            type = with types; nullOr (either str (listOf str));
```
What would a list of strings mean?

toonn replied:
I think I may have copied it blindly from `startAt`? Not sure, reading the systemd documentation it doesn't look like it would make sense. Good catch!

dotlambda reviewed (commented):
It would be nice to have a test case in which the server machine is started after the job is started for the first time in order to test whether the job is actually restarted.

toonn committed:
nixos/borgbackup: Fix documentation formatting

Co-authored-by: Robert Schütz <nix@dotlambda.de>

toonn committed:
nixos/borgbackup: Fix restartSec type

toonn commented:
@dotlambda, I have no experience with NixOS tests. How would I start a server after starting a job on it?

Stunkymonkey commented:
should this be closed in favor of #153346?

toonn commented:
Have to say this is leaving a rather sour taste in my mouth.

I'd also still like the Restart option so a failed attempt doesn't necessarily mean a lost backup.

Stunkymonkey commented:
@toonn could you please resolve the merge conflicts? I would be happy to have this merged. Please have a look at: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/backup/borgbackup.nix#L104=

toonn commented:
@Stunkymonkey, rebased.

I renamed `mkBacupTimers` to my original singular because the function never generates a config for more than one timer.

And I renamed `persistentTimer` to `persistent`, this is in line with other options, e.g. [nix.gc.persistent](https://nixos.org/manual/nixos/stable/options.html#opt-nix.gc.persistent), [system.autoupgrade.persistent](https://nixos.org/manual/nixos/stable/options.html#opt-system.autoUpgrade.persistent), [networking.dhcpcd.persistent](https://nixos.org/manual/nixos/stable/options.html#opt-networking.dhcpcd.persistent), and the `systemd.timers` terminology. I'm looking into how to generate a warning for this change now, so anyone who's set it knows to migrate.

toonn commented:
Found `mkRenamedOptionModule` but still looking into how that would work for an option with a generic `<name>` in the middle.

mweinelt commented:
> @dotlambda, I have no experience with NixOS tests. How would I start a server after starting a job on it?

By replacing `start_all()` with `client.start()` and trying to backup once while the server is still down. Then `server.start()` and wait for the retry.
